### PR TITLE
Add unit tests for JoinReturnWithAssignment.

### PR DESCRIPTION
**Pull request by scheglov**
This adds tests for patterns, as requested in https://github.com/dart-lang/linter/issues/4157

I also moved all other tests into the same unit tests class.

### CODE REVIEW

File: test/rules/join_return_with_assignment.dart [hunk 1]
```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
```
Comment (pq) [inline test/rules/join_return_with_assignment.dart:1]: Could you rename this file to `join_return_with_assignment_test.dart` and then add it to `all.dart` too?

**Review by pq — CHANGES_REQUESTED** [review]
Awesome!

File: test/rules/join_return_with_assignment.dart [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+import '../rule_test_support.dart';
+
+main() {
+  defineReflectiveSuite(() {
+    defineReflectiveTests(PreferFinalFieldsTest);
+  });
+}
+
+@reflectiveTest
+class PreferFinalFieldsTest extends LintRuleTest with LanguageVersion300Mixin {
```
[the remark this answers is not in the document]
Comment (scheglov) [reply test/rules/join_return_with_assignment.dart]: Done.

File: test/rules/join_return_with_assignment.dart (hunk 1 above)
Comment (scheglov) [reply test/rules/join_return_with_assignment.dart]: Done.

**Review by pq — APPROVED** [review]
Thanks!

**Pull request merged**